### DISCUSSION TITLE
custom directive quasar ssr

### DISCUSSION THREAD

**Discussion by murilolivorato** [quasarframework/quasar#10832] [CLI - SSR mode] [answered]
hello I have been trying to make a custom directive in quasar , I know that I must inicialize this directive in boot , like this -
https://quasar.dev/quasar-cli/developing-ssr/vue-ssr-directives
I tried to implement this directive on quasar but it was not successful - 
https://stackoverflow.com/questions/63869859/migrating-detect-click-outside-custom-directive-from-vue-2-to-vue-3
in quasar I showld start like this -

```
// src-ssr/directives/my-dir.js

export default (dir) => {
  return {
    props: []
  }
}
````

I dont understand , becose this directive start like this -

```
const { createApp } = Vue;
...
```
I dont knwo how could I place this code inside a directive of quasar boot .

**Comment by ontwikkelfabriek** [comment]
@murilolivorato 

Did you see this page in Quasar docs about custom directives, it comes with a full example custom directive( with it's own repo):
https://quasar.dev/app-extensions/tips-and-tricks/provide-a-directive#introduction

Also:
https://quasar.dev/quasar-cli/developing-ssr/vue-ssr-directives#how-to-declare-a-directive

**Reply by LefterisMavrakis** [reply]
Do we need to create an App extension in order to provide a custom directive for ssr builds?

**Reply by LefterisMavrakis** [reply]
I upgraded my quasar v2 app today and also vue3 and vue router 4. All of my directives doesnt working for ssr builds.

src/boot/click-outside-directive.js

```
`import { boot } from 'quasar/wrappers'

export default boot(async ({ app, router, store, ssrContext }) => {
  app.directive('click-outside', {
    updated(el, binding) {
      if(el.classList.contains('visible')) {
        el.clickOutsideEvent = function(event) {
          if(el !== event.target && !el.contains(event.target)) {
            binding.value();
          }
        }
        document.body.addEventListener('click', el.clickOutsideEvent);
        window.addEventListener('resize', () => {
          document.body.removeEventListener('click', el.clickOutsideEvent);
        })
      } else {
        document.body.removeEventListener('click', el.clickOutsideEvent);
      }
    },
    unmounted(el) {
      document.body.removeEventListener('click', el.clickOutsideEvent);
    },
    getSSRProps (binding) {
      // server-side implementation:
      // return the props to be rendered.
      // getSSRProps only receives the directive binding.
      return {
        class: 'click-outside'
      }
    }
  })
})`
```

src-ssr/directives/click-outside.js

`export default (clickOutside) => {
  return {
    props:[],
    getSSRProps(){
      return {}
    }
  }
}`

I still get
 [Vue warn]: Failed to resolve directive: click-outside
ssr error TypeError: Cannot read properties of undefined (reading 'getSSRProps')

Any ideas if there is something wrong to my implementation?

**Comment by murilolivorato** [comment]
hello @ontwikkelfabriek , thanks for answering .
I tried but it didn't work , could you help me ?
I am trying to make a simple directive like this example -
https://vueschool.io/lessons/custom-vuejs-directive-values
* I am using SSR .
in the example , it is like this -

```
<h1 v-color='purple'>this is a test</h1>
```
```
Vue.directive('purple', function(el) {
el.syle.color = 'purple'
})
new Vue({
   el: '#app',
   data: {}
})
```
in Quasar I did like this -

```
<h1 v-color='purple'>this is a test</h1>
```
My Boot -
```
export default ({ app }) => {
  app.directive('color', {
    mounted (el, binding) {
      el.style.color = 'purple'
    }
  })
}

```
But it didnt worked .

**Comment by hawkeye64** [comment]
l think your directive name should be "color" and not "purple" (otherwise the directive would ve `v-purple`)

**Reply by murilolivorato** [reply]
@hawkeye64, I did the change . It is working  only in SPA model .
It is not working in SSR Mode.

I tried as well to initialized the boot like this -

```
 boot: [
      { path: 'color', server: false }
    ],
``` 
But it didn't work , just work in SPA mode .

**Comment by hawkeye64** [comment]
That's because in SSR mode, there is no browser, so no `el`, just return,
You need to handle the `ssrContext`: https://quasar.dev/quasar-cli/developing-ssr/ssr-context#introduction

```js
export default ({ app, ssrContext }) => {
```

**Reply by murilolivorato** [reply]
@hawkeye64 , I don't know how could I do it . If I cant use el . 
how could I change the color style with directive in this case  ?
```
 app.directive('color', {
    mounted (el, binding) {
      el.style.color = 'purple'
    }
  })
```

showld I use ssrContext  instead ? something like that ?

```
 app.directive('color', {
    mounted (el, binding) {
      ssrContext.style.color = 'purple'
    }
  })

```

**Reply by hawkeye64** [reply]
Personally, I haven't done much with SSR. Does the `el` even get assigned something or is it `null` or `undefined`? I would think when running in SSR mode, just don't do anything, or check `el` for falsiness.

**Comment by murilolivorato** [comment]
I need help , I realy dont know how to do it .

**Comment by mohamedelhosseiny** [comment]
This is really frustrating. did anyone manage to find a solution for this?

**Comment by rstoenescu** [comment, marked as answer]
Commenting here for completeness.
Until recently, there was a bug in Vue that forced us to construct the src-ssr/directives approach. If you're on latest q/app-webpack or q/app-vite this is no longer needed (you can delete this folder as it is now ignored). You can define your directives directly in the UI (/src) and just make sure that you add getSSRProps() to each.
